### ISSUE TITLE
Slow integer-array indexing

### ISSUE THREAD

**Issue by SimonHeybrock** [scipp/scipp#3044]
https://scipp.github.io/user-guide/slicing.html#Integer-array-indexing is quite slow for long index arrays (considerably slower than NumPy). This may become more relevant when loading `NXoff_geometry` with ScippNexus, so we should look into this.

**Comment by jl-wynen** [comment]
Did some benchmarking of Scipp's and NumPy's functions. See the script below for how I did it.

NumPy's performance is an order of magnitude worse when indexing with a list compared to an array. In Scipp, there is no real difference. Both cases are an order of magnitude slower than the slow indexing by list in NumPy. (For n=2**9 in the script, 0.001 - 0.002 in Scipp and 0.0001 - 0.0002 in NumPy.)

## Bindings

I benchmarked the Pybind11 bindings by modifying the `__getitem__` implementation as follows. All times are for n = 2**9.

TL;DR the bindings with conversion to a `std::vector` alone take as long as NumPy's entire indexing routine. (When called with an array of indices.)

Empty body. Time: 0.00016 (index by list), 0.00028 (index by array)

```cpp
c.def(
    "__getitem__",
    [](T &self, const std::vector<scipp::index> &indices) {
      return;
    },
    py::call_guard<py::gil_scoped_release>());
```

Empty body + take indices as list. Time: 0.00004

```cpp
c.def(
    "__getitem__",
    [](T &self, const py::list &indices) {
      return;
    },
    py::call_guard<py::gil_scoped_release>());
```

Empty body + take indices as untyped array. Time: 0.00004

```cpp
c.def(
    "__getitem__",
    [](T &self, const py::array &indices) {
      return;
    },
    py::call_guard<py::gil_scoped_release>());
```

Empty body + take indices as typed array. Time: 0.00005

```cpp
c.def(
    "__getitem__",
    [](T &self, const py::array_t<scipp::index> &indices) {
      return;
    },
    py::call_guard<py::gil_scoped_release>());
```

## Using Python code: GIL

`slice_by_list` uses `py::slice` and therefore acquires the GIL many times in a loop. We can do away with it by implementing index computations ourselves. (Judging by #3102, the indexing with negative indices is broken / incomplete anyway.)

Modifying `slice_by_list` to use the following, I got a time of 0.0008, i.e. half way between the current release and NumPy with index lists.

```cpp
const auto make_slice = [](const scipp::index p, const scipp::index s) {
  return std::tuple{p, p + 1};
};
```

## The code

<details>

<summary>Benchmark script</summary>

```python
import timeit

from matplotlib import pyplot as plt
import numpy as np
import scipp as sc


def main():
    size = 1000
    array = np.random.uniform(0.0, 1.0, size)
    var = sc.array(dims=['x'], values=array)

    np_times = {}
    sc_times = {}
    for n in (2 ** e for e in (2, 3, 4, 5, 6, 7, 8, 9)):
        kwargs = {'stmt': 'x[i]',
                  'setup': 'i=(np.random.randint(0, size, size=n))',
                  'globals': {'np': np, 'size': size, 'n': n},
                  'number': 10,
                  'repeat': 100}

        kwargs['globals']['x'] = array
        times = timeit.repeat(**kwargs)
        np_times[n] = (min(times), max(times))

        kwargs['globals']['x'] = var
        times = timeit.repeat(**kwargs)
        sc_times[n] = (min(times), max(times))

    x = list(np_times)
    ynp_min, ynp_max = zip(*np_times.values())
    ysc_min, ysc_max = zip(*sc_times.values())

    plt.plot(x, ynp_min, label='np_min')
    plt.plot(x, ynp_max, label='np_max')
    plt.plot(x, ysc_min, label='sc_min')
    plt.plot(x, ysc_max, label='sc_max')
    plt.legend()
    plt.show()


if __name__ == '__main__':
    main()

```

</details>

**Comment by jl-wynen** [comment]
I'm struggling to reproduce my earlier speedup from avoiding the use of the GIL.

I tried a number of other improvements like taking a numpy array as argument and avoiding additional copies of the ranges by constructing an element_array directly instead of a vector.

It got me virtually no speedup over main. I did however notice that the construction of `to_extract` is quite expensive. Even in the form below. I don't understand why, I would have expected it to allocate one bit of memory, set up a ref counter and move a few things into place.

Here is my attempt:
```.cpp
template <class T>
T slice_by_list(const T &obj,
                const std::tuple<Dim, py::array_t<scipp::index>> &index) {
  const auto make_slice = [](scipp::index p, const scipp::index s) {
    if (p < 0)
        p = s + p;
    if (p < 0 || p>=s)
    throw std::out_of_range("The requested index " + std::to_string(p) +
                            " is out of range for dimension of length " +
                            std::to_string(s));
    return std::tuple{p, p+1};
  };

  // TODO check that indices are 1d
  const auto &[dim, index_array] = index;
  const scipp::span<const scipp::index> indices(index_array.data(), static_cast<size_t>(index_array.size()));
  const auto size = obj.dims()[dim];
  if (indices.size() != 0) {
    const auto [min, max] = std::minmax_element(indices.begin(), indices.end());
    if (*min < -size || *max >= size) {
      const auto bad = *min < -size ? *min : *max;
      throw_index_error(bad, size);
    }
  }
  core::element_array<scipp::index_pair> ranges(indices.size(), core::init_for_overwrite);
  auto *out_range = ranges.begin();
  for (const auto &pos : indices) {
    const auto [start, stop] = make_slice(pos, size);
    *out_range = scipp::index_pair{start, stop};
    ++out_range;
  }
  Variable to_extract(std::nullopt, Dimensions{{dim, indices.size()}}, std::move(ranges), std::optional<core::element_array<scipp::index_pair>>{});
  return extract_ranges(to_extract, obj, dim);
}
```
with binding (plus 2nd binding for multi-dim arrays)
```.cpp
  c.def(
      "__getitem__",
      [](T &self, py::array_t<scipp::index> indices) {
        expect_implicit_dimension(self.dims());
        return slice_by_list(self, {self.dim(), std::move(indices)});
      },
      py::call_guard<py::gil_scoped_release>());
```
